test/integration/functional-test.js

```diff
@@ -13,32 +13,261 @@ chai.use(chaiAsPromised);
+      await webext.inContent();
+      await helper.management();
+    });
+
+    beforeEach(async () => {
+      await clearLogins();
+    });
+
+    it("can add a new item", async () => {
+      const expected = {
+        origin: "https://foo.example.com",
+        username: "testUser",
+        password: "testPassword",
+      };
+
+      // Click the "+" button to add a new item.
+      const addButton = await waitFor(selectors.addItemButton);
+      expect(addButton).to.not.be.null;
+      await addButton.click();
+
+      // The new item form should be summoned.
+      const newItemForm = await waitFor(selectors.newItemForm);
+      expect(newItemForm).to.not.be.null;
+
+      // Grab the new item form fields.
+      const fields = {
+        origin: await waitFor(selectors.newItemFormField("origin")),
+        username: await waitFor(selectors.newItemFormField("username")),
+        password: await waitFor(selectors.newItemFormField("password")),
+      };
+
+      // Fill out the fields with expected item details.
+      for (let [name, field] of Object.entries(fields)) {
+        await field.sendKeys(expected[name]);
+      }
+
+      // Submit the fields.
+      const submitButton = await waitFor(selectors.newItemFormSubmit);
+      await submitButton.click();
+
+      // Verify visible Entry Details
+      const resultElements = {
+        origin: await waitFor(selectors.itemDetails("origin")),
+        username: await waitFor(selectors.itemDetails("username")),
+      };
+      for (let [name, el] of Object.entries(resultElements)) {
+        const text = await el.getText();
+        expect(text).to.equal(expected[name]);
+      }
+
+      // Verify visible info in the item list.
+      const listItemSubtitle = await waitFor(selectors.listItemSelectedSubtitle);
+      expect(await listItemSubtitle.getText()).to.equal(expected.username);
+
+      // Finally, ensure that our added item is found in Firefox
+      const listItemContainer = await waitFor(selectors.listItemContainer);
+      const itemId = await listItemContainer.getAttribute("data-item-id");
+      expect(
+        (await getLogins()).filter(login => login.guid == itemId).length,
+      ).to.equal(1);
+    });
+  });
+
+  describe("doorhanger", () => {
+    let popup;
+
+    before(async () => {
+      await webext.inContent();
+      popup = await helper.popup();
+    });
+
+    beforeEach(async () => {
+      await clearLogins();
+    });
+
+    it("lists logins", async () => {
+      // Add the login, then wait for "No results" to disappear.
+      await addLogin(mockLogin);
+      await waitUntilMissing(selectors.itemListEmpty);
+      await waitFor(selectors.itemList);
+
+      const listItemSubtitle = await waitFor(selectors.listItemSubtitle);
+      expect(await listItemSubtitle.getText()).to.equal(mockLogin.username);
+    });
+
+    it("displays login details on click", async () => {
+      // Add the login, then wait for "No results" to disappear.
```

Comment:
This is a really hacky feeling sequence, but it seems to be the one that didn't end up with intermittent failures for me :/

Reply:
I think this is a completely legit way of detecting DOM changes